In [1]:
import numpy as np
import pandas as pd
from pathlib import Path

In [5]:
train_df = pd.read_csv(Path('Ressources/2019loans.csv'))
test_df = pd.read_csv(Path('Ressources/2020Q1loans.csv'))

In [6]:
# Convert categorical data to numeric and separate target feature for training data
y_train = train_df.pop('loan_status')
X_train = pd.get_dummies(train_df)

print("X_train: {}, y_train: {}".format(X_train.shape, y_train.shape))

X_train: (12180, 94), y_train: (12180,)


In [7]:
# Convert categorical data to numeric and separate target feature for testing data
y_test = test_df.pop('loan_status')
X_test = pd.get_dummies(test_df)

print("X_test: {}, y_test: {}".format(X_test.shape, y_test.shape))

X_test: (4702, 93), y_test: (4702,)


In [14]:
# add missing dummy variables to testing set
missing_columns = set(X_train.columns).difference(X_test.columns)
print(missing_columns)

X_test["debt_settlement_flag_Y"] = 0

set()


In [16]:
# Train the Logistic Regression model on the unscaled data and print the model score
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import balanced_accuracy_score

logistic_regression_model = LogisticRegression(solver='lbfgs', random_state=1)
logistic_regression_model.fit(X_train, y_train)
y_pred = logistic_regression_model.predict(X_test)
balanced_accuracy_score(y_test, y_pred)

C:\Users\J-cle\anaconda3\envs\pythondata\lib\site-packages\sklearn\linear_model\_logistic.py:765: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


0.5253083794130158

In [21]:
# Train a Random Forest Classifier model and print the model score
from imblearn.ensemble import BalancedRandomForestClassifier
from sklearn.metrics import balanced_accuracy_score

random_forest_model = BalancedRandomForestClassifier(n_estimators=100, random_state=1)
random_forest_model.fit(X_train, y_train)
y_pred = random_forest_model.predict(X_test)
balanced_accuracy_score(y_test, y_pred)

0.6114419396001701

In [22]:
# Scale the data
from sklearn.preprocessing import StandardScaler

X_scaler = StandardScaler().fit(X_train)
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)
print("X_train_scaled: {}, X_test_scaled: {}".format(X_train_scaled.shape, X_test_scaled.shape))

X_train_scaled: (12180, 94), X_test_scaled: (4702, 94)


In [23]:
# Train the Logistic Regression model on the scaled data and print the model score
logistic_regression_model.fit(X_train_scaled, y_train)
y_pred = logistic_regression_model.predict(X_test_scaled)
balanced_accuracy_score(y_test, y_pred)

C:\Users\J-cle\anaconda3\envs\pythondata\lib\site-packages\sklearn\linear_model\_logistic.py:765: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


0.7205444491705657

In [24]:
# Train a Random Forest Classifier model on the scaled data and print the model score
random_forest_model.fit(X_train_scaled, y_train)
y_pred = random_forest_model.predict(X_test_scaled)
balanced_accuracy_score(y_test, y_pred)

0.6118672905146746

0.5253083794130158 Logistic Regression
0.6114419396001701 Balanced Random Forest Classifier
0.7205444491705657 Logistic Regression with Scaled: This one is the highest.
0.6118672905146746 Random Foresst with Scaling